controllers/azurejson_machinepool_controller_test.go

```diff
@@ -210,3 +217,140 @@ func TestAzureJSONPoolReconciler(t *testing.T) {
+	g.Expect(err).ToNot(HaveOccurred())
+
+	err = clusterv1.AddToScheme(scheme)
+	g.Expect(err).ToNot(HaveOccurred())
+
+	err = infrav1.AddToScheme(scheme)
+	g.Expect(err).ToNot(HaveOccurred())
+
+	azureMP := &infrav1exp.AzureMachinePool{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "fake-machine-pool",
+			Namespace: "fake-ns",
+			Labels: map[string]string{
+				clusterv1.ClusterNameLabel: "fake-cluster",
+			},
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: expv1.GroupVersion.Group + "/" + expv1.GroupVersion.Version,
+					Kind:       "MachinePool",
+					Name:       "fake-other-machine-pool",
+					Controller: to.Ptr(true),
+				},
+			},
+		},
+		Spec: infrav1exp.AzureMachinePoolSpec{
+			UserAssignedIdentities: []infrav1.UserAssignedIdentity{
+				{
+					ProviderID: "fake-id",
+				},
+			},
+		},
+	}
+
+	clusterA := &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "fake-cluster",
+			Namespace: "fake-ns",
+		},
+		Spec: clusterv1.ClusterSpec{
+			InfrastructureRef: &corev1.ObjectReference{
+				Kind:      "AzureCluster",
+				Name:      "fake-azure-cluster",
+				Namespace: "fake-ns",
+			},
+		},
+	}
+
+	ownerMP := &expv1.MachinePool{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "fake-other-machine-pool",
+			Namespace: "fake-ns",
+			Labels: map[string]string{
+				clusterv1.ClusterNameLabel: "fake-cluster",
+			},
+		},
+	}
+
+	azureCluster := &infrav1.AzureCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "fake-azure-cluster",
+			Namespace: "fake-ns",
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: "cluster.x-k8s.io/v1beta1",
+					Kind:       "Cluster",
+					Name:       "my-cluster",
+				},
+			},
+		},
+		Spec: infrav1.AzureClusterSpec{
+			AzureClusterClassSpec: infrav1.AzureClusterClassSpec{
+				SubscriptionID: "123",
+			},
+			NetworkSpec: infrav1.NetworkSpec{
+				Subnets: infrav1.Subnets{
+					{
+						SubnetClassSpec: infrav1.SubnetClassSpec{
+							Name: "node",
+							Role: infrav1.SubnetNode,
+						},
+					},
+				},
+			},
+		},
+	}
+	apiVersion, kind := infrav1.GroupVersion.WithKind("AzureMachinePool").ToAPIVersionAndKind()
+
+	sec := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      azureMP.Name,
+			Namespace: "fake-ns",
+			Labels: map[string]string{
+				"fake-cluster": string(infrav1.ResourceLifecycleOwned),
+			},
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: apiVersion,
+					Kind:       kind,
+					Name:       azureMP.GetName(),
+					Controller: ptr.To(true),
+				},
+			},
+		},
+	}
+
+	client := fake.NewClientBuilder().WithScheme(scheme).WithObjects(azureMP, ownerMP, clusterA, azureCluster, sec).Build()
+	rec := AzureJSONMachinePoolReconciler{
+		Client:           client,
+		Recorder:         record.NewFakeRecorder(42),
+		ReconcileTimeout: 20 * time.Second,
+	}
+	id := "fake-id"
+	getClient = func(auth azure.Authorizer) (identities.Client, error) {
```

Comment:
Where is this function used?

Reply:
`getClient` equals `identities.NewClient` 

If we don't patch that we have to pass an authorizer into the `identities.NewClient` func and try to auth to Azure.  Even if we pass an mocked authorizer, it will still try to go out and make the call to Aure to provide the "identity" client.  

This allows us to override this method during testing but not during production.  

In the test we do this 
```go
 getClient = func(auth azure.Authorizer) (identities.Client, error) {
       mockClient := mock_identities.NewMockClient(ctrlr)
      mockClient.EXPECT().GetClientID(gomock.Any(), gomock.Any()).Return(id, nil)
      return mockClient, nil
     }
```
So when we call `Reconcile, it doesn't try to go out and make a call to Azure but use our mock_identities caller in this method.`

So basically "patching" the method so we can test or else we get blocked here from a unit perspective.